### ISSUE TITLE
Files should not have colons in their filename

### ISSUE THREAD

**Issue by amyworrall** [apple/swift-docc#284]
DocC currently generates filenames containing the Swift method names of the code that is being documented. These frequently contain the colon character.

This character is disallowed on Windows. Therefore, we should generate filenames with a different character in its place, and ensure all links/URLs throughout the generated documentation respect the new convention.

### Expected behavior
All characters in filenames generated by DocC are valid characters on major filesystems used by Windows, MacOS and Linux.

### Actual behavior
Filenames contain colons, which are not valid on Windows.

### Steps to Reproduce
Generate DocC documentation, using transform for static hosting. Observe the filenames of files that document Swift methods that contain at least one parameter.

### Swift-DocC Version Information

**Swift-DocC version:** 
The version shipped with Xcode 13.4.0

**Swift Compiler version:**
swift-driver version: 1.45.2 Apple Swift version 5.6.1 (swiftlang-5.6.0.323.66 clang-1316.0.20.12)
Target: x86_64-apple-macosx12.0

**Comment by ethan-kusters** [comment]
Thank you for filing this @amyworrall! Some additional context here: https://forums.swift.org/t/docc-colons-in-filenames/57917.

**Comment by compnerd** [comment]
@ethan-kusters - this is now blocking progress on supporting Windows.  What would be the best place to perform the substitution?  `FileSafeURL`?  I am tempted with going with the simplistic approach of replacing invalid characters (`[<>:"/|?*\\]`) with `_`.

**Comment by ethan-kusters** [comment]
CC: @d-ronnqvist.

I think `FileSafeURL` is likely the right place but David has more context here. How are you thinking of implementing this? It's likely worth starting a new forums thread so we can discuss further.

I can imagine a handful of approaches:

1. Just enable this new mode when building on Windows.
2. Create a new flag that can be passed to `docc convert` to enable this new mode. (and maybe also enable it by default on Windows)
3. Make this the new way all DocC web urls are created – regardless of build/server platform

My concerns here are generally the same as before - I think the Web URLs are much more readable with the colons so I'd personally rather leave this as opt-in and likely enabled by default on Windows for local development builds. But it would be good to get the input of the community here.

**Comment by d-ronnqvist** [comment]
I don't know anything about the Windows file system or how a web server on Windows handles requests but my main question to determine the right change would be:
- is it only the file URL that need to change or does the path of the web URL also need to change so that the server can route the request and load the expected file?

If it's only the file URL that need to change then `FileSafeURL` is probably a good place for it. This is likely going to be small in scope.

If the web URL also need to change then there's both more places in the code that would need to be updated and there are new questions about the user experience that would need to be discussed in the Swift Forums. Readability of the web URLs is an important factor but there's also many thousands of documentation pages that have colons in their URLs today that would break—both to system that link directly to their web URL and to systems that inspect the DocC "linkable entities" file.

**Comment by ethan-kusters** [comment]
> If it's only the file URL that need to change then FileSafeURL is probably a good place for it. This is likely going to be small in scope.

I _think_ changing the file URL would also affect the web URL at least for deployments relying on static file server hosting.

**Comment by compnerd** [comment]
@d-ronnqvist yes, the url path itself would have to change as the : would no longer be in the file name and the web server needs to route the request accordingly.

@ethan-kusters while I understand that it's easier url to read, having the additional path is something I would prefer to not have.  Perhaps an alternative would be to opt out of the unsafe character substitution by a flag?
